### ISSUE TITLE
Separate input rules required for IPSec decapsulation

### ISSUE THREAD

**Issue by pprindeville** [firewalld/firewalld#680]
I'm running into a limitation in `Firewalld` where I don't want to accept cleartext GRE directly on a public interface, but I would accept it if it were encapsulated in IPSec/ESP.

Suggest having a 2nd zone that gets run post decapsulation so that separate access control lists are available on IPSec tunneling.

**Comment by erig0** [comment]
Sounds like what you actually want is support for matching ipsec fields (exists, spi, etc). Then you could do something like:
```
# firewall-cmd --add-rich-rule='rule family=ipv4 ipsec exists protocol value=gre accept'
```
which would allow GRE packets that are also ipsec.

**Comment by pprindeville** [comment]
Either as a related issue or not, while I was searching for a possible workaround I noticed that:

- while `richlanguage` marking as an action, it doesn't support matching on a mark as a condition;
- you also can't match on the equivalent of `-m policy --pol ipsec --dir in`;
- and you can't define or jump to an additional chain.

these might be useful enhancements to `firewalld`.

**Comment by erig0** [comment]
> Either as a related issue or not, while I was searching for a possible workaround I noticed that:
> 
>     * while `richlanguage` marking as an action, it doesn't support matching on a mark as a condition;

Correct. This would be a useful enhancement. You can file a separate RFE if you'd like.

>     * you also can't match on the equivalent of `-m policy --pol ipsec --dir in`;

Right. Which is what I think is being proposed in this RFE.

>     * and you can't define or jump to an additional chain.

That's not something intended by rich rules. If you want that kind of flexibility then maybe you should just directly use `nftables` or `iptables`.

**Comment by pprindeville** [comment]
> Right. Which is what I think is being proposed in this RFE.

Well, yes, and no.  This RFE is strictly suggesting a separate chain be zone be added to handle post-decapsulation traffic.  So if you have a zone `public`, then you'd also have a zone `public_decaps` (for instance) that gets invoked once the re-injected packet gets processed.

**Comment by pprindeville** [comment]
> Correct. This would be a useful enhancement. You can file a separate RFE if you'd like.

Done as #681.

**Comment by erig0** [comment]
> > Right. Which is what I think is being proposed in this RFE.
> 
> Well, yes, and no. This RFE is strictly suggesting a separate zone be added to handle post-decapsulation traffic. So if you have a zone `public`, then you'd also have a zone `public_decaps` (for instance) that gets invoked once the re-injected packet gets processed. Hence the original title.

Well if that's the case then it's a firm WONTFIX from me. Packets should ingress one and only one zone. This is a tenant of all zone based firewalls.

**Comment by pprindeville** [comment]
> Well if that's the case then it's a firm WONTFIX from me. Packets should ingress one and only one zone. This is a tenant of all zone based firewalls.

Then how would you address this issue?

**Comment by erig0** [comment]
> > Well if that's the case then it's a firm WONTFIX from me. Packets should ingress one and only one zone. This is a tenant of all zone based firewalls.
> 
> Then how would you address this issue?

As I said in my first comment. Enhance rich rules to allow matching ipsec.

**Comment by pprindeville** [comment]
I'm trying to find the sweet spot between minimalism (i.e. _lightweight design_) and out-of-the-box utility so that your average user can secure his network without needing to become a firewall expert, that is having helpful canned rules for common usage scenarios.

**Comment by erig0** [comment]
Development note:

In nftables to match incoming ipsec packets (already decapsulated) we can use `ipsec exists`. To match outgoing packets that will be encapsulated we can use `rt ipsec`. Further support can be use `ipset {in|out} ..`, but that's probably not needed for most use cases.
Encapsulated packets can be matched by matching `esp` or `ah` protocols.

**Comment by pprindeville** [comment]
> I'm trying to find the sweet spot between minimalism (i.e. lightweight design) and out-of-the-box utility so that your average user can secure his network without needing to become a firewall expert, that is having helpful canned rules for common usage scenarios.

To put this comment in better perspective, a _router_-based firewall is typically administered by a network engineer or network administrator who has understanding of network protocols, or can find an read a tutorial on that, and typically only has to secure a perimeter firewall.

A _host_-based firewall, on the other hand, is (1) much more prevalent since there are usually many more hosts than networks, and (2) managed by a system administrator who might have limited networking experience.

My perspective (as a protocol engineer and security architect) is that a host firewall that requires too much knowledge to ramp up and use effectively is self-defeating.

I think that as VPN becomes more the norm (especially in this era of COVID and so many people working from home), it's actually a fairly common scenario to integrate VPN (and in particular, IPSec-based VPN) into off-the-shelf configurations.

**Comment by erig0** [comment]
The ipsec packets will pass through netfilter twice.

1. encapsulated (encrypted payload)
    - ingress interface is the "uplink", e.g. eth0
    - at this point you can match the packet using protocols: esp, ah
2. after decapsulation (decrypted payload)
    - ingress interface is the ipsec interfaces, e.g. ipsec0
    - you can use a dedicated zone for this
    ```
    # firewall-cmd --permanent --new-zone ipsec
    # firewall-cmd --permanent --zone --add-interface ipsec0
    ```

Then add your rules to the `ipsec` zone as you would for any other zone.
The rich rule match I suggested in this RFE is useful if you want to combine ipsec packet processing with a normal zone.

**Comment by pprindeville** [comment]
Conversation on `#Netfilter` suggests that convention wisdom is that each zone should have its own post-processing zone for IPSec, but since in my current (urgent) scenario I'll only ever be receiving packets from `public`, I can live with this for now.

**Comment by pprindeville** [comment]
If we made things sufficiently extensible, it should be possible to also add a post-GRE decapsulation hook and pass things through a zone that way...

**Comment by erig0** [comment]
> If we made things sufficiently extensible, it should be possible to also add a post-GRE decapsulation hook and pass things through a (new) zone that way...

You do exactly as I said above. Create another zone add use `--add-source gre0`.
